corehq/apps/enterprise/tasks.py

```diff
@@ -92,3 +92,101 @@ def auto_deactivate_mobile_workers():
         task_time.seconds,
         multiprocess_mode=MPM_LIVESUM
     )
+
+
+@task(queue="export_download_queue")
+def generate_enterprise_report(slug, billing_account_id, username, **kwargs):
+    report = EnterpriseReport.create(slug, billing_account_id, username, **kwargs)
+
+    rows = report.rows
+    progress = ReportTaskProgress(slug, username)
+    progress.complete_task(rows)
+
+
+class TaskProgress:
+    """
+    This class manages a task's lifecycle. Tasks triggered by this class are expected to call
+    complete_task in order to store their results for later retrieval.
+    Once a task completes, a query_id is generated that can be used to retrieve the results in the future.
+    Multiple tasks run with the same key will overwrite each other's results. If you want to re-use a key after
+    it has completed its task, you can call clear_status so that new tasks will store their results
+    in a different query_id
+    """
+    STATUS_TIMEOUT = 600  # 10 minutes
+
+    STATUS_COMPLETE = 'COMPLETE'
+    STATUS_IN_PROGRESS = 'IN_PROGRESS'
+    STATUS_NEW = 'NEW'
+
+    def __init__(self, key, query_id=None):
+        self.key = key
+        self._query_id = query_id
+        self.redis_client = get_redis_client()
+
+    def get_status(self):
+        if self._query_id:
+            return self.STATUS_COMPLETE
+
+        status_dict = self.redis_client.get(self.key)
+        if not status_dict:
+            return self.STATUS_NEW
+
+        return status_dict['status']
+
+    def start_task(self, task):
+        status_dict = {
+            'status': self.STATUS_IN_PROGRESS,
+            'query_id': None,
+            'task': task,
+        }
+
+        self.redis_client.set(self.key, status_dict, timeout=self.STATUS_TIMEOUT)
+        task.delay()
+
+    def complete_task(self, data):
+        if not self.get_query_id():
+            self._query_id = uuid.uuid4().hex
```

Comment:
In the specific case you mention, `complete_task` would raise an exception, because the status entry would have been cleared by `clear_status`. While a race condition could exist, the code doesn't currently use it in a way where this could happen. If a caller wanted to use it in that way, then I think it's fair for them to guard against those possibilities with synchronization constructs. Is the second part of this comment suggesting your previous implementation? Accepting a `query_id` parameter comes with its own downsides -- namely, it means that a client can query for any task's results. I pinged @esoergel to ask for a second opinion, but I've read through the implementation and am happy with the current iteration

Reply:
> In the specific case you mention, complete_task would raise an exception

Good point. I hadn't noticed that. Would it make sense to fetch the status record before storing the data value in redis since it will never be used if the status record does not exist?

> Is the second part of this comment suggesting your previous implementation?

No.

Synchronization constructs?? Why are you insisting on keeping code that is unused and has a very complicated usage protocol?

This code is untested, and I have spent hours reviewing it on multiple occasions. I have found it difficult to understand, and have tried unsuccessfully ask for simplifications. I have nothing more to say here.